rp2: Allow changing the clock frequency.

robert-hh opened the pull request:
Using machine.freq(). The safe ranges tested were 10 and 12-270MHz,
at which USB REPL still worked.
Requested settings can be checked with the script:
pico-sdk/src/rp2_common/hardware_clocks/scripts/vcocalc.py
At frequencies like 300MHz the script still signaled OK, but USB
did not work any more.

robert-hh committed:
modmachine.c: Allow changing clock frequency.

Using machine.freq(). The safe ranges tested were 10 and 12-270MHz,
at which USB REPL still worked.
Requested settings can be checked with the script:
pico-sdk/src/rp2_common/hardware_clocks/scripts/vcocalc.py
At frequencies like 300MHz the script still signaled OK, but USB
did not work any more.

stinos commented on ports/rp2/modmachine.c:
```diff
@@ -80,10 +82,19 @@ STATIC mp_obj_t machine_bootloader(void) {
+        return MP_OBJ_NEW_SMALL_INT(clock_get_hz(clk_sys));
+    } else {
+        mp_int_t freq = mp_obj_get_int(args[0]);
+        if (set_sys_clock_khz(freq / 1000, false)) {
```
Bit more concise and clear (at least to me):
```
if (!set_sys_clock_khz(freq / 1000, false)) {
    mp_raise_ValueError(MP_ERROR_TEXT("valid range 12-270MHz"));
}
return mp_const_none;
``` 
?

robert-hh replied:
I prefer to  show both wings or a if/else statement.

robert-hh replied:
But I changed it. Don't care about it.

stinos commented on ports/rp2/modmachine.c:
```diff
@@ -80,10 +82,19 @@ STATIC mp_obj_t machine_bootloader(void) {
+        if (set_sys_clock_khz(freq / 1000, false)) {
+            return mp_const_none;
+        } else {
+            mp_raise_ValueError(MP_ERROR_TEXT("valid range 12-270MHz"));
```
I get that this is nor completely arbitrary, on the other hand there doesn't seem to be a guarantee that these numbers actually work, nor that numbers out if this range won't, so seems better to just state "cannot change frequency"?

robert-hh replied:
My first message was something like "invalid frequency". But then I thought could give people a hint, after trying a handful values. But testing now the range for the full MHz values, 45 values of that range are not accepted. So a more vague text is better. I'll change that. Since it's a ValueError and the call has just a single argument, it might not need a message at all. Because that's what it is.

stinos replied:
All odd numbers i.e. clock divider cannot do that? That mechanism is probably documented somewhere..

robert-hh replied:
Not all odd numbers. 141, 147, 153, 159, 165,  ... are OK, and everything between 12 and 136 is fine too, as well as 10.

robert-hh committed:
modmachine.c: Changed error message & line order.

The error message was less helpful than initially considered, since
in the range of 12-270 MHz a total of 45 integer MHZ values were not
accepted, not even to speak of fractional values.

robert-hh commented:
I always fail on the commit message format by not adding a full stop to the header. I read it as a title, and title do not end on a full stop.

stinos commented:
Yes, I have the same problem but got used to it. Also if you think about reading it as "This commit will <message here>." then the full stop makes more sense.

robert-hh committed:
modmachine.c: Re-init UART for REPL on freq change.

When UART is for REPL and the MCU frequency is changed, the UART
has to be re-initialised. Besides that UART may have to be recreated
after frequency change, but with USB REPL this is not a problem.

Thanks to @HermannSW for spotting and providing the change.

dpgeorge commented:
Thanks, this is a good addition.  Merged in 11cf742524202fa5fc062f3e6e3040a82f49b190 and c675452566ffcae1a49121b84a0233ed164b2152
